File: firefly/src/main/java/com/firefly/server/http2/router/impl/PrecisePathMatcher.java
package com.firefly.server.http2.router.impl;

import com.firefly.server.http2.router.Matcher;
import com.firefly.server.http2.router.Router;

import java.util.*;

/**
 * @author Pengtao Qiu
 */
public class PrecisePathMatcher implements Matcher {

    private Map<String, Set<Router>> precisePath;

    private Map<String, Set<Router>> precisePath() {
        if (precisePath == null) {
            precisePath = new HashMap<>();
        }
        return precisePath;
    }

    @Override
    public void add(String rule, Router router) {
        precisePath().computeIfAbsent(rule, k -> new HashSet<>()).add(router);
    }

    @Override
    public MatchResult match(String value) {
        if (precisePath == null) {
            return null;
        }

        Set<Router> routers = precisePath.get(value);
        if (routers != null) {
            return new MatchResult(routers, Collections.emptyMap(), MatchType.PATH);
        } else {
            return null;
        }
    }
}

File: firefly-net/src/main/kotlin/com/fireflysource/net/http/server/impl/matcher/PrecisePathMatcher.kt
package com.fireflysource.net.http.server.impl.matcher

import com.fireflysource.net.http.server.Matcher
import com.fireflysource.net.http.server.Matcher.MatchType
import com.fireflysource.net.http.server.Router

class PrecisePathMatcher : AbstractPreciseMatcher() {

    override fun add(rule: String, router: Router) {
        val path = toPath(rule)
        super.add(path, router)
    }

    override fun match(value: String): Matcher.MatchResult? {
        if (routersMap.isEmpty()) return null

        val path = toPath(value)
        return super.match(path)
    }

    override fun getMatchType(): MatchType? {
        return MatchType.PATH
    }

    private fun toPath(value: String): String = if (value.last() != '/') "$value/" else value
}